### PR TITLE
Add filters for transactions and their events

### PR DESCRIPTION
**Pull request by nebolax**
Part of #4220

Added ability to filter transactions by related events on endpoint `/blockchains/ETH/transactions` with `GET` method.

No changelog since this feature also requires frontend implementation.

### CODE REVIEW

File: rotkehlchen/api/rest.py [hunk 1]
```diff
@@ -3029,18 +3028,19 @@ def _get_ethereum_transactions(
...
             database=self.rotkehlchen.data.db,
         )
-        transactions: Optional[List[EthereumTransaction]]
         try:
             transactions, total_filter_count = tx_module.query(
                 only_cache=only_cache,
                 filter_query=filter_query,
                 has_premium=self.rotkehlchen.premium is not None,
             )
+            transactions = list(set(transactions))
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:3043]: Why are you doing this here?

File: rotkehlchen/db/filtering.py [hunk 2]
```diff
@@ -103,6 +103,22 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
         return filters, bindings
 
 
+@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=False)
+class DBETHTransactionEventsJoin(DBFilter):
+    """Adds history_events to the query, so they can be used for filtering"""
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:108]:
```suggestion
    """Adds history_events to the transactions query, so they can be used for filtering"""
```

File: rotkehlchen/db/filtering.py [hunk 3]
```diff
@@ -295,6 +311,9 @@ def make(
             from_ts: Optional[Timestamp] = None,
             to_ts: Optional[Timestamp] = None,
             tx_hash: Optional[EVMTxHash] = None,
+            filter_by_events: Optional[bool] = None,  # If True, filters by events data
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:314]: Two things here.

Why give this argument? Why not just filter by events data if either the `protocol` or the `asset`argument is not `None`?

File: rotkehlchen/db/filtering.py [hunk 4]
```diff
@@ -312,6 +331,12 @@ def make(
         else:
             if addresses is not None:
                 filter_query.join_clause = DBETHTransactionAddressFilter(and_op=False, addresses=addresses)  # noqa: E501
+            elif filter_by_events:
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:334]: Hmm so what happens if we want to filter both by address and either asset and/or protocol?

File: rotkehlchen/db/filtering.py [hunk 5]
```diff
@@ -673,6 +698,7 @@ def make(
             ignored_ids: Optional[List[str]] = None,
             null_columns: Optional[List[str]] = None,
             event_identifier: Optional[str] = None,
+            counterparty: Optional[str] = None,
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:701]: I would make this both here and in the given api a list of counterparties. There is no need to filter only by one protocol. User should be able to filter by multiple protocols.

File: rotkehlchen/tests/api/test_ethereum_transactions.py [hunk 6]
```diff
@@ -1051,3 +1067,143 @@ def query_transactions(rotki, tx_module: EthTransactions) -> None:
     ):
         assert cursor.execute(f'SELECT COUNT(*) from {name}').fetchone()[0] == 0
     assert dbevents.get_history_events(HistoryEventFilterQuery.make(), True) == []
+
+
+def create_tx(tx_hash: bytes) -> EthereumTransaction:
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_ethereum_transactions.py:1072]: You know what would be nice? Add this as `make_ethereum_transaction()` here: https://github.com/rotki/rotki/blob/develop/rotkehlchen/tests/utils/factories.py

And you can even have the tx_hash as optional argument. If not given you can make it create a random tx hash

File: rotkehlchen/tests/api/test_ethereum_transactions.py [hunk 7]
```diff
@@ -1051,3 +1067,143 @@ def query_transactions(rotki, tx_module: EthTransactions) -> None:
...
+        value=1,
+        gas=1,
+        gas_price=1,
+        gas_used=1,
+        input_data=b'',
+        nonce=0,
+    )
+
+
+def create_tx_event(
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_ethereum_transactions.py:1089]: Same thing for this. `make_tx_event`

File: rotkehlchen/tests/api/test_ethereum_transactions.py [hunk 8]
```diff
@@ -1051,3 +1067,143 @@ def query_transactions(rotki, tx_module: EthTransactions) -> None:
...
+    dbevents.add_history_events([event1, event2, event3, event4])
+
+    response = requests.get(
+        api_url_for(
+            rotkehlchen_api_server,
+            'ethereumtransactionsresource',
+        ),
+        json={
+            'asset': A_ETH.serialize(),
+        },
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_ethereum_transactions.py:1162]: fits in one line

```suggestion
        json={'asset': A_ETH.serialize()},
```

File: rotkehlchen/tests/api/test_ethereum_transactions.py [hunk 9]
```diff
@@ -1051,3 +1067,143 @@ def query_transactions(rotki, tx_module: EthTransactions) -> None:
...
+    assert result['entries'] == expected
+
+    response = requests.get(
+        api_url_for(
+            rotkehlchen_api_server,
+            'ethereumtransactionsresource',
+        ),
+        json={
+            'asset': A_BTC.serialize(),
+        },
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_ethereum_transactions.py:1175]: fits in one line

```suggestion
        json={'asset': A_BTC.serialize()},
```

File: rotkehlchen/tests/api/test_ethereum_transactions.py [hunk 10]
```diff
@@ -1051,3 +1067,143 @@ def query_transactions(rotki, tx_module: EthTransactions) -> None:
...
+    assert result['entries'] == expected or result['entries'] == list(reversed(expected))
+
+    response = requests.get(
+        api_url_for(
+            rotkehlchen_api_server,
+            'ethereumtransactionsresource',
+        ),
+        json={
+            'protocol': 'EXAMPLE_PROTOCOL',
+        },
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_ethereum_transactions.py:1191]: fits in one line

```suggestion
        json={'protocol': 'EXAMPLE_PROTOCOL'},
```

File: rotkehlchen/tests/api/test_ethereum_transactions.py [hunk 11]
```diff
@@ -1051,3 +1067,143 @@ def query_transactions(rotki, tx_module: EthTransactions) -> None:
...
+        json={
+            'asset': A_BTC.serialize(),
+        },
+    )
+    result = assert_proper_response_with_result(response)
+    expected = generate_tx_entries_response([(tx1, [event3]), (tx2, [event4])])
+    # For some reason data this data can be reversed,
+    # and we avoid failing with a help of this ugly check.
+    # Dicts are not hashable, so it's not possible to use better and simpler way
+    assert result['entries'] == expected or result['entries'] == list(reversed(expected))
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_ethereum_transactions.py:1182]: I don't really understand what you are doing here. Why is the result not deterministic here? It should be either one or the other

**Review by LefterisJP — COMMENTED** [review]
Some comments / questions

File: rotkehlchen/db/filtering.py (hunk 4 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:334]: This is a question that we with @yabirgb spent time on and came up with this solution. There can be only one `join_clause` and we decided that it will be too much to alter filter classes structure.

File: rotkehlchen/api/rest.py (hunk 1 above)
Comment (nebolax) [reply rotkehlchen/api/rest.py:3043]: Hm... Looks like I've forgotten about `DISTINCT` keyword in SQL :))

File: rotkehlchen/db/filtering.py (hunk 2 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:108]: Alright

File: rotkehlchen/db/filtering.py (hunk 3 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:314]: Yes, could be removed

File: rotkehlchen/tests/api/test_ethereum_transactions.py (hunk 7 above)
Comment (nebolax) [reply rotkehlchen/tests/api/test_ethereum_transactions.py:1089]: Cool

File: rotkehlchen/db/filtering.py (hunk 4 above)
Comment (yabirgb) [reply rotkehlchen/db/filtering.py:334]: You can use the location label field of the history events. For eth transacctions is the user address

File: rotkehlchen/api/v1/schemas.py [hunk 12]
```diff
@@ -188,19 +190,29 @@ def make_ethereum_transaction_query(  # pylint: disable=no-self-use
     ) -> Dict[str, Any]:
         address = data.get('address')
         order_by_attribute = data['order_by_attribute'] if data['order_by_attribute'] is not None else 'timestamp'  # noqa: E501
+        protocols, asset = data['protocols'], data['asset']
+        if not protocols:  # If protocols == []
+            protocols = None
```
Comment (yabirgb) [inline rotkehlchen/api/v1/schemas.py:195]: I believe lefteris will agree that here is better to be explicit and why to write it in the comment when you can write it as code. I would use 
```suggestion
        if len(protocols) == 0:
            protocols = None
```

This would prevent (maybe not here but in other scenarios) possible errors that come from invalid types for protocols

File: rotkehlchen/db/filtering.py [hunk 13]
```diff
@@ -84,23 +84,33 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
-            f'A.tx_hash=b.tx_hash AND B.address IN ({",".join(questionmarks)})',
-        ]
-        bindings = self.addresses
+        query_filter, bindings = '', []
+        if self.join_events:
+            query_filter += str(
+                ' INNER JOIN '
+                '(SELECT event_identifier, counterparty, asset FROM history_events)'
+                'ON hex(ethereum_transactions.tx_hash)=substr(event_identifier, 3)',
+            )
```
Comment (yabirgb) [inline rotkehlchen/db/filtering.py:103]:
```suggestion
            query_filter += (
                ' INNER JOIN '
                '(SELECT event_identifier, counterparty, asset FROM history_events)'
                'ON hex(ethereum_transactions.tx_hash)=substr(event_identifier, 3)'
            )
```

**Review by yabirgb — COMMENTED** [review]
2 small comments

File: rotkehlchen/api/v1/schemas.py (hunk 12 above)
Comment (nebolax) [reply rotkehlchen/api/v1/schemas.py:195]: I used `not protocols` here because `protocols` can be either `None` or `[]` so `len(protocols)` might fail. But I surely can change it to a bigger explicit condition like `if protocols is None or len(protocols) == 0`

File: rotkehlchen/api/v1/schemas.py [hunk 14]
```diff
@@ -188,19 +190,29 @@ def make_ethereum_transaction_query(  # pylint: disable=no-self-use
     ) -> Dict[str, Any]:
         address = data.get('address')
         order_by_attribute = data['order_by_attribute'] if data['order_by_attribute'] is not None else 'timestamp'  # noqa: E501
+        protocols, asset = data['protocols'], data['asset']
+        if protocols is None or len(protocols) == 0:
+            protocols = None
```
Comment (yabirgb) [inline rotkehlchen/api/v1/schemas.py:195]: But with this condition this you are reassigning `protocols=None`. What I would have written is
```suggestion
        if protocols is not None and len(protocols) == 0:
            protocols = None
```

**Review by yabirgb — COMMENTED** [review]
Small comment to improve the code

File: rotkehlchen/api/v1/schemas.py (hunk 14 above)
Comment (nebolax) [reply rotkehlchen/api/v1/schemas.py:195]: Alright

File: rotkehlchen/db/ethtx.py [hunk 15]
```diff
@@ -156,10 +156,10 @@ def get_ethereum_transactions(
         cursor = self.db.conn.cursor()
         query, bindings = filter_.prepare()
         if has_premium:
-            query = 'SELECT * FROM ethereum_transactions ' + query
+            query = 'SELECT DISTINCT ethereum_transactions.tx_hash, timestamp, block_number, from_address, to_address, value, gas, gas_price, gas_used, input_data, nonce FROM ethereum_transactions ' + query  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/db/ethtx.py:159]: why are you making this change here and below?

Why `DISTINCT` and why spell out all fields?

File: rotkehlchen/db/filtering.py [hunk 16]
```diff
@@ -84,23 +84,33 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
 @dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=False)
-class DBETHTransactionAddressFilter(DBFilter):
+class DBETHTransactionJoinsFilter(DBFilter):
     """Find transactions involving any of the addresses. Including in an internal.
 
     This uses the mappings we create in the DB at transaction addition to signify
     relevant addresses for a transaction.
     """
-    addresses: List[ChecksumEthAddress]
+    addresses: Optional[List[ChecksumEthAddress]]
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:93]: Also update docstring then explaining the different uses of this filter and what do the combinations mean. What is join_events etc.

File: rotkehlchen/db/filtering.py [hunk 17]
```diff
@@ -84,23 +84,33 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
+        query_filter, bindings = '', []
+        if self.join_events:
+            query_filter += (
+                ' INNER JOIN '
+                '(SELECT event_identifier, counterparty, asset FROM history_events)'
+                'ON hex(ethereum_transactions.tx_hash)=substr(event_identifier, 3)'
+            )
+        if self.addresses:
+            questionmarks = '?' * len(self.addresses)
+            query_filter += str(
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:106]: Why cast to str?

```suggestion
            query_filter += (
```

File: rotkehlchen/db/filtering.py [hunk 18]
```diff
@@ -84,23 +84,33 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
+                f'INNER JOIN ethtx_address_mappings ON '
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:107]: This seems broken. If both `self.join_events` and `self.addresses` have truthy values you end up with a broken string: `'...(event_identifier, 3)INNER JOIN ...'`

Please fix this, and think about this a bit. This is the reason we have filters in a list and we join them later. Like this the code is super brittle and easy to misunderstand and break.

What's more this means that all the combinations are definitely not tested in this PR. So please create unit tests of all filter combinations of this one.

File: rotkehlchen/db/filtering.py [hunk 19]
```diff
@@ -84,23 +84,33 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
 
     def prepare(self) -> Tuple[List[str], List[Any]]:
-        questionmarks = '?' * len(self.addresses)
-        filters = [
-            f'AS A LEFT OUTER JOIN ethtx_address_mappings AS B WHERE '
-            f'A.tx_hash=b.tx_hash AND B.address IN ({",".join(questionmarks)})',
-        ]
-        bindings = self.addresses
+        query_filter, bindings = '', []
+        if self.join_events:
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:98]:
```suggestion
        if self.join_events is True:
```

File: rotkehlchen/db/filtering.py [hunk 20]
```diff
@@ -84,23 +84,33 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
+            query_filter += (
+                ' INNER JOIN '
+                '(SELECT event_identifier, counterparty, asset FROM history_events)'
+                'ON hex(ethereum_transactions.tx_hash)=substr(event_identifier, 3)'
+            )
+        if self.addresses:
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:104]: explicit checks of what is a truthy value

```suggestion
        if self.addresses is not None:
```

File: rotkehlchen/db/filtering.py [hunk 21]
```diff
@@ -310,8 +322,17 @@ def make(
         if tx_hash is not None:  # tx_hash means single result so make it as single filter
             filters.append(DBETHTransactionHashFilter(and_op=False, tx_hash=tx_hash))
         else:
-            if addresses is not None:
-                filter_query.join_clause = DBETHTransactionAddressFilter(and_op=False, addresses=addresses)  # noqa: E501
+            if addresses or asset or protocols:
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:325]: Explicit truthy checks

```suggestion
            if addresses is not None or asset is not None or protocols is not None:
```

File: rotkehlchen/db/filtering.py [hunk 22]
```diff
@@ -310,8 +322,17 @@ def make(
...
+                filter_query.join_clause = DBETHTransactionJoinsFilter(
+                    and_op=False,
+                    addresses=addresses,
+                    join_events=bool(asset or protocols),
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:329]: you already have this check above. So why not keep it in a variable and use it here too without redoing it?

File: rotkehlchen/db/filtering.py [hunk 23]
```diff
@@ -744,3 +770,14 @@ def make(
         filters.append(filter_query.timestamp_filter)
         filter_query.filters = filters
         return filter_query
+
+
+@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=False)
+class DBProtocolsFilter(DBFilter):
+    protocols: List[str]
+
+    def prepare(self) -> Tuple[List[str], List[Any]]:
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:779]: This will fail for empty list of protocols. But if we stop empty list of protocols at the api level it should be impossible.

So please write a comment here specifying that `protocols` can't be empty since it's checked by flask (implement said check and test it)

File: rotkehlchen/tests/api/test_ethereum_transactions.py (hunk 6 above)
Comment (LefterisJP) [reply rotkehlchen/tests/api/test_ethereum_transactions.py:1061]: I don't see you having done it :)

File: rotkehlchen/tests/api/test_ethereum_transactions.py [hunk 24]
```diff
@@ -1051,3 +1056,108 @@ def query_transactions(rotki, tx_module: EthTransactions) -> None:
...
+    assert result['entries'] == expected
+
+    response = requests.get(
+        api_url_for(
+            rotkehlchen_api_server,
+            'ethereumtransactionsresource',
+        ),
+        json={
+            'asset': A_BTC.serialize(),
+            'protocols': ['EXAMPLE_PROTOCOL'],
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_ethereum_transactions.py:1158]: Also add a test for frontend giving to the backend an empty list of protocols and see that the backend reacts properly. I would even say empty list of protocols makes no sense. So reject it at the api level please.

**Review by LefterisJP — COMMENTED** [review]
lots of stuff

File: docs/api.rst [hunk 25]
```diff
@@ -1894,7 +1894,7 @@ Querying ethereum transactions
    .. note::
       This endpoint can also be queried asynchronously by using ``"async_query": true``
 
-   Doing a GET on the transactions endpoint for ETH will query all ethereum transactions for all the tracked user addresses. Caller can also specify an address to further filter the query as a from address. Also they can limit the queried transactions by timestamps. If the user is not premium and has more than 500 transaction then the returned transaction will be limited to that number. Any filtering will also be limited to those first 500 transaction. Transactions are returned most recent first.
+   Doing a GET on the transactions endpoint for ETH will query all ethereum transactions for all the tracked user addresses. Caller can also specify an address to further filter the query as a from address. Also they can limit the queried transactions by timestamps and can filter transactions by related event's properties (asset and protocol). If the user is not premium and has more than 500 transaction then the returned transaction will be limited to that number. Any filtering will also be limited to those first 500 transaction. Transactions are returned most recent first.
```
[the remark this answers is not in the document]
Comment (nebolax) [reply docs/api.rst:1897]: Alright

File: rotkehlchen/db/filtering.py (hunk 16 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:93]: Yep, updated

File: rotkehlchen/db/filtering.py (hunk 18 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:107]: Yes, I understand, will improve tests and fix the query.

Filters can have only one possible `join_clause` property so we with @yabirgb found this as the best way to do multiple joins.

File: rotkehlchen/db/filtering.py (hunk 19 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:98]: Alright

File: rotkehlchen/db/filtering.py (hunk 20 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:104]: But what if addresses is `[]`? In this case the query won't return anything. Is it okay?

File: rotkehlchen/db/filtering.py (hunk 22 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:329]: Yep, good idea

File: rotkehlchen/tests/api/test_ethereum_transactions.py (hunk 24 above)
Comment (nebolax) [reply rotkehlchen/tests/api/test_ethereum_transactions.py:1158]: Aha. Alright, will add logic to reject it

Comment (nebolax) [reply rotkehlchen/tests/api/test_ethereum_transactions.py:1158]: Added

File: rotkehlchen/db/filtering.py (hunk 23 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:779]: Added check, test and a comment

File: rotkehlchen/db/filtering.py (hunk 20 above)
Comment (LefterisJP) [reply rotkehlchen/db/filtering.py:108]: Well I would say an empty addresses list to filter should never make it down here. We should prevent it at api level.

But semantically if you tell me to filter for a list of addresses and that list is empty the query should indeed return nothing

File: rotkehlchen/db/filtering.py [hunk 26]
```diff
@@ -310,8 +326,18 @@ def make(
         if tx_hash is not None:  # tx_hash means single result so make it as single filter
             filters.append(DBETHTransactionHashFilter(and_op=False, tx_hash=tx_hash))
         else:
-            if addresses is not None:
-                filter_query.join_clause = DBETHTransactionAddressFilter(and_op=False, addresses=addresses)  # noqa: E501
+            join_events = bool(asset is not None or protocols is not None)
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:329]: Why do you need to cast this to bool?

File: rotkehlchen/db/filtering.py [hunk 27]
```diff
@@ -310,8 +326,18 @@ def make(
...
+            if addresses is not None or join_events:
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:330]:
```suggestion
            if addresses is not None or join_events is True:
```

File: rotkehlchen/db/ethtx.py (hunk 15 above)
Comment (LefterisJP) [reply rotkehlchen/db/ethtx.py:159]: Hmm okay would have to see how this plays in practise. Thanks

File: rotkehlchen/db/filtering.py [hunk 28]
```diff
@@ -84,23 +84,37 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
-        filters = [
-            f'AS A LEFT OUTER JOIN ethtx_address_mappings AS B WHERE '
-            f'A.tx_hash=b.tx_hash AND B.address IN ({",".join(questionmarks)})',
-        ]
-        bindings = self.addresses
+        query_filter, bindings = '', []
+        if self.join_events is True:
+            query_filter += (
+                ' INNER JOIN '
+                '(SELECT event_identifier, counterparty, asset FROM history_events)'
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:105]:
```suggestion
                '(SELECT event_identifier, counterparty, asset FROM history_events) '
```

**Review by LefterisJP — COMMENTED** [review]
Left some comments and questions. Also still not sure how multiple join clauses will work but perhaps we can see that in subsequent PRs

File: rotkehlchen/db/filtering.py (hunk 20 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:108]: Alright!

File: rotkehlchen/db/filtering.py (hunk 27 above)
Comment (nebolax) [reply rotkehlchen/db/filtering.py:330]: Yep

**Review by LefterisJP — APPROVED** [review]
LGTM
